File: apps/mobile/android/app/src/main/java/xyz/speakeasyapp/app/voicefilter/ActiveFilterHolder.kt
package xyz.speakeasyapp.app.voicefilter

/**
 * Process-wide singleton holding the currently-active voice filter
 * for capture audio. Read on every WebRTC audio frame from the
 * forked [org.webrtc.audio.WebRtcAudioRecord], written by
 * [VoiceFilterModule] when JS calls `wrapTrack` / `dispose`.
 *
 * The static singleton is intentional: the device has one mic, the
 * WebRTC `JavaAudioDeviceModule` is per-`PeerConnectionFactory` (and
 * react-native-webrtc creates one factory per app), and the brand
 * promise is that filtering is on or off at the moment audio is
 * captured. Threading a per-track filter through `WebRtcAudioRecord`
 * would require forking more of WebRTC's surface for no real
 * benefit — there's only ever one active call at a time.
 *
 * Volatile because the WebRTC audio-record thread reads it without
 * any synchronization; we accept the (rare) race where a call ends
 * while a frame is mid-flight by checking the returned `process`
 * boolean and muting the mic when the filter is gone or has tripped.
 */
object ActiveFilterHolder {
  @Volatile private var current: SampleFilter? = null

  /** Read by the WebRTC audio-record thread. May return null. */
  @JvmStatic
  fun getFilter(): SampleFilter? = current

  /** Install [filter] (or null to detach). Called by [VoiceFilterModule]. */
  @JvmStatic
  fun setFilter(filter: SampleFilter?) {
    current = filter
  }
}

File: apps/mobile/ios/SpeakeasyBridges/VoiceFilter/ActiveFilterHolder.swift
//
//  ActiveFilterHolder.swift
//  Speakeasy — Phase 5j Private Call voice filter
//
//  Process-wide holder for the currently-active voice filter on
//  capture audio. Mirrors the Android-side
//  `xyz.speakeasyapp.app.voicefilter.ActiveFilterHolder` contract:
//  the audio device reads on every capture frame, VoiceFilterModule
//  writes when JS calls `wrapTrack` / `dispose`.
//
//  Atomic via an internal serial queue. Reads from the audio thread
//  are lock-free except during the brief swap window.
//

import Foundation

final class ActiveFilterHolder {
  static let shared = ActiveFilterHolder()
  private init() {}

  /// `os_unfair_lock` would be lighter, but for one-pointer load on
  /// the audio thread the cost is dominated by AVAudioEngine's own
  /// queue overhead. A serial queue is plenty.
  private let lock = NSLock()
  private var _current: SampleFilter?

  func get() -> SampleFilter? {
    lock.lock()
    defer { lock.unlock() }
    return _current
  }

  func set(_ filter: SampleFilter?) {
    lock.lock()
    _current = filter
    lock.unlock()
  }
}
